flipperdevices/flipperzero-firmware
# [FL-3070] iButton system and app refactoring

# What's new

- Refactored iButton subsystem for easier addition of new key types
- Refactored iButton app (new gui and behaviour)
- Added support for:
  * DS1992 (with 1 kbit SRAM)
  * DS1996 (with 64 kbit SRAM) keys (no overdrive yet)
- Any unknown 1-Wire devices show up as "Generic"

# Verification 

1. Download [iButtonTestFiles.zip](https://github.com/flipperdevices/flipperzero-firmware/files/10850838/iButtonTestFiles.zip), extract it and put the files to the `/ext/ibutton` directory on the Flipper Zero.
2. Go to `Saved` menu. Ensure that all of the above files can be loaded without any errors. Both old v1 formatted files and new v2 files must work.
3. After selecting a saved file, choose `Emulate` menu option. Repeat it for all of the above files. (Use another Flipper to read the DS1992 and DS1996 keys fully. The latter will take several seconds to read as there is no overdrive support yet). Ensure that all of the files can be emulated without errors.
4. For all keys with the `Write blank` option available, select it and try writing the key onto a blank. Read the blank back to ensure that it has been correctly written.
5. For all keys with the `Write copy` option available, select it and try writing the key onto a key of the same type (i.e. a DS1992 onto another DS1992). Read the written key back to ensure there were no errors.
6. Go to `Read` mode and try reading all of the available iButton key types. All keys must be read without errors.
7. For all keys with the `View data` option, select it and ensure that the data was correctly read.
8. Go to `Add Manually` menu and add one key of each protocol. For all Dallas protocols (except DSGeneric), a correct family code and checksum must be pre-filled. The checksum must update itself each time the user changes the data. Save the file and load it to ensure that the file was formed properly.
9. Go to `Saved -> Edit` and edit the key. Save the file and load it to ensure that the changes were saved properly.

# Checklist (For Reviewer)
